### PR TITLE
Jy/using mango

### PR DESCRIPTION
**Pull request by jackiep00**
serum-history has a really good event deduplicator that we've used

### CODE REVIEW

File: src/events.ts [hunk 1]
```diff
@@ -33,7 +46,7 @@ const EVENT = struct([
 
 export function decodeRecentEvents(buffer: Buffer, lastSeenSeqNum?: number) {
   const header = EVENT_QUEUE_HEADER.decode(buffer);
-  const nodes: any[] = [];
+  const events: any[] = [];
```
Comment (peebeejay) [inline src/events.ts]: could be typed

File: src/index.ts [hunk 2]
```diff
@@ -150,7 +150,18 @@ const main = async function () {
       console.log(marketMeta['name']);
 
       let loadTimestamp = new Date().toISOString();
-      let events: FullEvent[] = await market.loadFillsAndContext(connection, 1000);
+      // let events: FullEvent[] = await market.loadFillsAndContext(connection, 1000);
```
Comment (peebeejay) [inline src/index.ts]: remove?

File: src/index.ts [hunk 3]
```diff
@@ -102,11 +102,11 @@ function safeWrite(
...
-    if (!writeSuccess) {
-      // Writestream got overloaded!
-      // Drain and write some more once it drains
-      writer.once('drain', write);
-    }
+    // if (!writeSuccess) {
+    //   // Writestream got overloaded!
+    //   // Drain and write some more once it drains
+    //   writer.once('drain', write);
+    // }
```
Comment (peebeejay) [inline src/index.ts]: remove?

File: package.json [hunk 4]
```diff
@@ -27,9 +27,11 @@
   "dependencies": {
     "@project-serum/serum": "^0.13.38",
     "@solana/web3.js": "^1.17.0",
+    "@types/node-cron": "^2.0.3",
```
Comment (peebeejay) [inline package.json]: just fyi - types should ideally go under dev dependencies

File: src/uploader.ts [hunk 5]
```diff
@@ -2,15 +2,75 @@ import { statSync, createReadStream } from 'fs';
...
+    const stats = statSync(this.fileName, { throwIfNoEntry: false });
+    // returns undefined if the file does not exist
+    const fileSizeMB = stats ? stats.size / (1024 * 1024) : 0;
+
+    if (fileSizeMB > this.batchFilesizeMB) {
+      let loadTimestamp = new Date().toISOString();
+      const uploadFile = this.fileName;
+      // update the fileName before attempting the upload so we don't block the scrapers
+      this.fileName = `${this.fileNameTemplate}${loadTimestamp}.csv`;
+      console.log(
```
Comment (peebeejay) [inline src/uploader.ts:52]: we should probably remove this `console.log` and any others that are still lurking around. 

should ideally be replaced w logger statements instead

File: src/uploader.ts [hunk 6]
```diff
@@ -2,15 +2,75 @@ import { statSync, createReadStream } from 'fs';
...
+  async batchUploadtoS3(): Promise<void> {
+    logger.info(`Checking filesize of ${this.fileName}`);
+    const stats = statSync(this.fileName, { throwIfNoEntry: false });
+    // returns undefined if the file does not exist
+    const fileSizeMB = stats ? stats.size / (1024 * 1024) : 0;
+
+    if (fileSizeMB > this.batchFilesizeMB) {
+      let loadTimestamp = new Date().toISOString();
+      const uploadFile = this.fileName;
+      // update the fileName before attempting the upload so we don't block the scrapers
```
Comment (peebeejay) [inline src/uploader.ts:50]: i think it would be good to include more context / info on why the scrapers are blocked here!

File: src/uploader.ts [hunk 7]
```diff
@@ -2,15 +2,75 @@ import { statSync, createReadStream } from 'fs';
...
+    this.ACL = ACL;
+    this.batchFilesizeMB = batchFilesizeMB;
+  }
+
+  // this function checks the file to see if it's larger than the threshold before actually uploading
+  // the filename it returns is current working file
+  async batchUploadtoS3(): Promise<void> {
+    logger.info(`Checking filesize of ${this.fileName}`);
+    const stats = statSync(this.fileName, { throwIfNoEntry: false });
+    // returns undefined if the file does not exist
```
Comment (peebeejay) [inline src/uploader.ts:44]: comments in the imperative instead? nbd lol

File: src/scraper.ts [hunk 8]
```diff
@@ -4,20 +4,23 @@ import { appendFile, writeFile, readFile, access, createWriteStream } from 'fs';
...
   }
 
   async getParseSaveEventsAsync(connection: Connection) {
     const loadTimestamp = new Date().toISOString();
     let marketMeta = { ...this.targetMarket };
 
-    let fileName = `output/all_market_events_${loadTimestamp}.csv`;
-    createWriteStream(fileName);
+    let fileName = this.uploader.fileName;
+    // createWriteStream(fileName);
```
Comment (peebeejay) [inline src/scraper.ts:23]: we should remove in a follow up PR!

**Review by peebeejay — APPROVED** [review]
just a few minor comments we can jointly address in a followup PR. looks sick

File: src/uploader.ts (hunk 6 above)
Comment (jackiep00) [reply src/uploader.ts:50]: this is inaccurate actually - a holdover from a failed attempt to solve the problem before i just serialized everything

Comment (peebeejay) [reply src/uploader.ts:50]: ah ok - lemme remove. writing a quick follow up

**Pull request merged**